### ISSUE TITLE
MethodLengthChecker should not count comments

### ISSUE THREAD

**Issue by lordgordon** [scalastyle/scalastyle#166]
Hi there,

I just noticed that `MethodLengthChecker` is raised independently of the nature of the lines: I have a method with 23 lines of actual code, yet `MethodLengthChecker` warning is raised. 
In fact my method is longer than 50 lines thanks to multiple comments.

It's possible to have this warning raised only counting the effective code, ignoring commented lines (ie lines starting with // or inside a multiline comment)?

Thanks

**Comment by matthewfarwell** [comment]
If you want to change the method length checker to add an option to ignore comment-only lines, then please go ahead. I don't want to change the behaviour by default though.

**Comment by shishkin** [comment]
@lordgordon I agree that ignoring comments shouldn't be the default behavior of the method length checker. The goal of the checker is raise an alarm when a method grows to big and to complex to comprehend.

If the comments are there to aid comprehension, extract portions of the method into methods of their own with meaningful names and remove comments.

If on the other hand the code is commented out, it's already dead. Just delete it, it's still in the VCS anyway.

Does it make sense?

**Comment by lordgordon** [comment]
I agree that changing the default behavior is not a good idea and this could be a new optional feature.

Maybe it's just me, I tend to write a lot of comments, also to add references to actual bug tracking issues to remember why a line of code was added. You know, on big project maintain legacy code is not easy.

I also noticed that empty lines are also counted in the length of a method (sometimes I use empty lines to make the code more readable, to separate).

BTW, of course I'm not talking about commented dead code :)

**Comment by shishkin** [comment]
I know how difficult it is to maintain large legacy code bases. Inflating the code base with comments even more is hardly of any help.

Git commit messages might be a valid alternative to comments for bug tracking references. It is less distracting and is still accessible if needed.

Separating sections of a method with blank lines is an indication of multiple responsibilities of the method. Sections should give you a hint, which portions to extract to methods of their own.

In the end finding other ways to express intent without comments leads to cleaner code.

**Comment by jawshooah** [comment]
@shishkin I agree with you that gratuitous comments can add more confusion than they remove in the long run, and that refactoring for clarity should always be preferred.

However, it seems to me that a foundational goal of Scalastyle is to not be overly opinionated, and to offer as much flexibility as possible in defining style requirements.

While ignoring comments in line counts should absolutely not be a default, it should be an option.

**Comment by shishkin** [comment]
@jawshooah Thanks. I didn't know ScalaStyle is aimed to not be opinionated. I though it's the other way around ;) Then of course, it makes sense to accommodate this as an optional feature.

**Comment by jawshooah** [comment]
Check out the full [list of rules](http://www.scalastyle.org/rules-0.7.0.html). You'll note that many come in pairs, allowing you to enforce or forbid a given practice depending on which rule you enable. I feel that this implies a lack of opinionation; that is, there is no inherent preference for one choice or the other.

There is, of course, a default configuration, but that's based on the developers' own style preferences.
